src/java.base/share/classes/java/lang/reflect/code/op/CoreOp.java

```diff
@@ -1396,125 +1396,168 @@ public sealed interface ReflectiveOp {
+                        case InvokeKind k -> k;
+                        case null, default -> {
+                            if (isVarArgs) {
+                                throw new UnsupportedOperationException("Unsupported invoke kind value:" + v);
+                            }
+                            int paramCount = invokeDescriptor.type().parameterTypes().size();
+                            int argCount = def.operands().size();
+                            yield (argCount == paramCount + 1)
+                                    ? InvokeKind.INSTANCE
+                                    : InvokeKind.STATIC;
+                        }
+                    });
+
+
+            return new InvokeOp(def, ik, isVarArgs, invokeDescriptor);
+        }
+
+        InvokeOp(ExternalizedOp def, InvokeKind invokeKind, boolean isVarArgs, MethodRef invokeDescriptor) {
             super(def);
 
+            validateArgCount(invokeKind, isVarArgs, invokeDescriptor, def.operands());
+
+            this.invokeKind = invokeKind;
+            this.isVarArgs = isVarArgs;
             this.invokeDescriptor = invokeDescriptor;
             this.resultType = def.resultType();
         }
 
         InvokeOp(InvokeOp that, CopyContext cc) {
             super(that, cc);
 
+            this.invokeKind = that.invokeKind;
+            this.isVarArgs = that.isVarArgs;
             this.invokeDescriptor = that.invokeDescriptor;
             this.resultType = that.resultType;
         }
 
-        InvokeOp(String name, TypeElement resultType, MethodRef invokeDescriptor, List<Value> args) {
-            super(name, args);
+        @Override
+        public InvokeOp transform(CopyContext cc, OpTransformer ot) {
+            return new InvokeOp(this, cc);
+        }
+
+        InvokeOp(InvokeKind invokeKind, boolean isVarArgs, TypeElement resultType, MethodRef invokeDescriptor, List<Value> args) {
+            super(NAME, args);
+
+            validateArgCount(invokeKind, isVarArgs, invokeDescriptor, args);
 
+            this.invokeKind = invokeKind;
+            this.isVarArgs = isVarArgs;
             this.invokeDescriptor = invokeDescriptor;
             this.resultType = resultType;
         }
 
+        static void validateArgCount(InvokeKind invokeKind, boolean isVarArgs, MethodRef invokeDescriptor, List<Value> operands) {
+            int paramCount = invokeDescriptor.type().parameterTypes().size();
+            int argCount = operands.size() - (invokeKind == InvokeKind.STATIC ? 0 : 1);
+            if ((!isVarArgs && argCount != paramCount)
+                    || argCount < paramCount - 1) {
+                throw new IllegalArgumentException(invokeKind + " " + isVarArgs + " " + invokeDescriptor);
+            }
+        }
+
         @Override
         public Map<String, Object> attributes() {
             HashMap<String, Object> m = new HashMap<>(super.attributes());
             m.put("", invokeDescriptor);
+            if (isVarArgs) {
```

Comment:
We don't need to in all cases - only required for super invocations. I just chose a path of least resistance when hacking it together. My main motivation was to avoid updating all the tests! Varargs calls are less common so I was not overly concerned about the scary strings.

But, for consistency I think it is worth addressing this. If `invoke.varargs` is present and `invoke.kind` is not we can infer static or instance invocation.

Reply:
Actually there was a reason, that i forgot about. Let's say we have a method `A::m(A... more)` we know it is varargs but don't know if it's a static or instance method. Given one operand we cannot differentiate between `a.m()` and `m(a)`.

Reply:
ugh - ok :-)